STM32Cube/Tasks/trajectory.c

```diff
@@ -0,0 +1,256 @@
+    * @param h time step
+    * @param extension extension of airbrakes, 0-1
+    * @param mass of rocket (kg)
+    * @param state, including altitude (m) and velocity in X and Y directions (m/s)
+    * @return updated altitude and velocity integrals after one rk4 step
+*/
+RK4State rk4(float h, float mass, float extension, RK4State state) {
+    Forces forces;
+
+    forces = get_forces(extension, mass, state.velX, state.velY, state.alt);
+    float ka1 = h * state.velY;
+    float kvY1 = h * acceleration(forces.Fy, mass);
+    float kvX1 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX1/2, state.velY + kvY1/2, state.alt + ka1/2);
+    float ka2 = h * (state.velY + h*kvY1/2);
+    float kvY2 = h * acceleration(forces.Fy, mass);
+    float kvX2 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX2/2, state.velY + kvY2/2, state.alt + ka2/2);
+    float ka3 = h * (state.velY + h*kvY2/2);
+    float kvY3 = h * acceleration(forces.Fy, mass);
+    float kvX3 = h * acceleration(forces.Fx, mass);
+
+    forces = get_forces(extension, mass, state.velX + kvX3, state.velY + kvY3, state.alt + ka3);
+    float ka4 = h * (state.velY + h*kvY3);
+    float kvY4 = h * acceleration(forces.Fy, mass);
+    float kvX4 = h * acceleration(forces.Fx, mass);
+
+    RK4State updatedState;
+    updatedState.alt = (state.alt + (ka1 + 2*ka2 + 2*ka3 + ka4)/6);
+    updatedState.velY = (state.velY + (kvY1 + 2*kvY2 + 2*kvY3 + kvY4)/6);
+    updatedState.velX = (state.velX + (kvX1 + 2*kvX2 + 2*kvX3 + kvX4)/6);
+
+    return updatedState;
+}
+
+/** @return max apogee
+* @param velocity vertical velocity (m/s)
+* @param altitude (m)
+* @param airbrake_ext extension of airbrakes, 0-1
+* @param mass (kg)
+*/
+float get_max_altitude(float velY, float velX, float altitude, float airbrake_ext, float mass) {
+
+    const float h = 0.05; // interval of change for rk4
+    float prevAlt = 0.0; // variable to store previous altitude
+        
+    RK4State states;
+    
+    states.alt = altitude;
+    states.velY = velY;
+    states.velX = velX;
+
+    while (states.alt >= prevAlt) {
+        prevAlt = states.alt; //to check if altitude is decreasing to exit the loop
+        states = rk4(h, mass, airbrake_ext, states);  // update velocity and altitude
+
+    }
+
+    return prevAlt;
+}
+
+void trajectory_task(){    
+    altQueue = xQueueCreate(1, sizeof(AltTime));
+    angleQueue = xQueueCreate(1, sizeof(AnglesUnion));
+    apogeeQueue = xQueueCreate(1, sizeof(float));
+    extQueue = xQueueCreate(1, sizeof(float));
+
+    float prev_time = -1;
+    uint16_t prev_alt = -1;
+    for(;;)
+    {
+        AltTime altTime;
+        AnglesUnion angles;
+        float ext;
+        if(xQueueReceive(altQueue, &altTime, 10) == pdTRUE) {
+            if(xQueueReceive(angleQueue, &angles, 10) == pdTRUE) {
+                if(xQueueReceive(extQueue, &ext, 10) == pdTRUE) {
```

Comment:
this is not correct, first of all I the behaviour we are going for is if the altitude has a new value from barometer (new value in queue) then we will update the rk4, regardless of if there is new values in the other two queues (there will be those go much faster but still). Additionally, right now, if there is a reading in the altitude qeueu, so we recive it, but there is not an updated value in on of the other two, then we remove that element from the altitude qeueue, and the next time that upates say in 100ms, it is very likely we would have recived new values for the other two, which run on more of a 10ms basis. But now since the queue is empty we wont act, until the next reading (and then hope the other two queues have data else around we go again). Regardless because of the intended behaviour, we only actually need to implement a recive for the first qeuee for alitutde since it will be driving the update of rk4, the other two qeueues can simply be peeked for their value. The thread upating theese qeueues can use xQueueOverwrite since the qeueue will be of size on (essentially representing the most up to date value we have for that item, basically just a memory safe global varible).

Reply:
This may be a communication error by me. But what I told jacob is essentially that state estimation will run so fast that we are guaranteed to always have an orientation value by the time we see a new altitude value. the controller extension is more of an issue as its very to easy to get into a deadlock scenario (and i believe if you just run this, it will in fact default to that - controller will not run until it gets an apogee, and prediction will not run until it has an extension.

So. Need to make sure that we peek the extension and angle queues, and need to make sure that we overwrite the value in the apogee queue. the timeout on the altitude queue can also be arbitrarily long, but I'm expecting altitudes on the order of 100 or even 1000ms.

I will deal with the interface with CAN and the other tasks